File: server/game/vec2.go
package game

import (
	"math"
)

type Vec2 struct {
	X float64
	Y float64
}

func (v Vec2) Add(other Vec2) Vec2 {
	return Vec2{
		X: v.X + other.X,
		Y: v.Y + other.Y,
	}
}

func (v Vec2) Sub(other Vec2) Vec2 {
	return Vec2{
		X: v.X - other.X,
		Y: v.Y - other.Y,
	}
}

func (v Vec2) Scale(s float64) Vec2 {
	return Vec2{
		X: v.X * s,
		Y: v.Y * s,
	}
}

func (v Vec2) Normalize() Vec2 {
	mag := v.Magnitude()
	if mag == 0.0 {
		return Vec2{}
	}

	return Vec2{
		X: v.X / mag,
		Y: v.Y / mag,
	}
}

func (v Vec2) Magnitude() float64 {
	return math.Sqrt(v.MagnitudeSq())
}

func (v Vec2) MagnitudeSq() float64 {
	return v.X*v.X + v.Y*v.Y
}

func (v Vec2) DistanceTo(other Vec2) float64 {
	temp := other.Sub(v)
	return temp.Magnitude()
}

func (v Vec2) Lerp(other Vec2, t float64) Vec2 {
	return Vec2{
		X: v.X * t + other.X * (1.0 - t),
		Y: v.Y * t + other.Y * (1.0 - t),
	}
}

File: client/scripts/vec2.js
export class Vec2 {
    /**
     * @param {number} x
     * @param {number} y
     */
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }

    /**
     * @param {Vec2} other
     */
    add(other) {
        return new Vec2(
            this.x + other.x,
            this.y + other.y
        );
    }

    /**
     * @param {Vec2} other
     */
    sub(other) {
        return new Vec2(
            this.x - other.x,
            this.y - other.y
        );
    }

    /**
     * @param {number} s
     */
    scale(s) {
        return new Vec2(
            this.x * s,
            this.y * s
        );
    }

    normalize() {
        let mag = this.magnitude();
        if (mag == 0.0) {
            return new Vec2(0, 0);
        }

        return new Vec2(
            this.x / mag,
            this.y / mag
        );
    }

    magnitude() {
        return Math.sqrt(this.x * this.x + this.y * this.y);
    }

    magnitudeSq() {
        return this.x * this.x + this.y * this.y;
    }

    /**
     * @param {Vec2} other
     */
    distanceTo(other) {
        let temp = other.sub(this);
        return temp.magnitude();
    }

    /**
     * @param {Vec2}   other
     * @param {number} t
     */
    lerp(other, t) {
        return new Vec2(
            this.x * t + other.x * (1.0 - t),
            this.y * t + other.y * (1.0 - t),
        )
    }
}
